libstuff/SScheduledPriorityQueue.h

```diff
@@ -0,0 +1,315 @@
+    typedef int Priority;
+    typedef uint64_t Timeout; 
+    typedef uint64_t Scheduled;
+
+    // If nothing becomes available to dequeue while waiting, a timeout_error exception is thrown.
+    class timeout_error : exception {
+      public:
+        const char* what() const noexcept {
+            return "timeout";
+        }
+    };
+
+    // By default, the start and end functions are No-ops.
+    SScheduledPriorityQueue(function<void(T& item)> startFunction = [](T& item){},
+                            function<void(T& item)> endFunction = [](T& item){})
+      : _startFunction(startFunction), _endFunction(endFunction) {};
+
+    // Remove all items from the queue.
+    void clear();
+
+    // Returns true if there are no queued commands.
+    bool empty();
+
+    // Returns the size of the queue.
+    size_t size();
+
+    // Get an item from the queue. Optionally, a timeout can be specified.
+    // If timeout is non-zero, a timeout_error exception will be thrown after waitUS microseconds, if no work was
+    // available.
+
+    // Get an item from the queue. If `waitUS` is specified, and no items become available before the specified time
+    // period has elapsed, timeout_error is thrown.
+    T get(uint64_t waitUS = 0);
+
+    // Same as above, but increments a counter just before returning an item. This allows callers to keep accurate
+    // counts of items in or out of the queue.
+    T get(uint64_t waitUS, atomic<int>& incrementBeforeDequeue);
+
+    // Add an item to the queue. The queue takes ownership of the item and the caller's copy is invalidated.
+    void push(T&& item, Priority priority, Scheduled scheduled, Timeout timeout);
+
+  protected:
+
+    // Associate the item with it's timeout so that when we dequeue an item to return, we can also remove it's entry
+    // in our set of timeouts.
+    struct ItemTimeoutPair {
+        ItemTimeoutPair(T&& _item, Timeout _timeout) : item(move(_item)), timeout(_timeout) {}
+        T item;
+        Timeout timeout;
+    };
+
+    // Removes an item from the queue and returns it, if a suitable item is available. Throws `out_of_range` otherwise.
+    T _dequeue(atomic<int>& incrementBeforeDequeue);
+
+    // Synchronization primitives for managing access to the queue.
+    mutex _queueMutex;
+    condition_variable _queueCondition;
+
+    // The main queue is a map of priorities to the items queued at that priority, sorted by their scheduled time.
+    map<Priority, multimap<Scheduled, ItemTimeoutPair>> _queue;
+
+    // A map of timeouts back into the respective priority queue to find the item with the given timeout.
+    multimap<Timeout, pair<Priority, Scheduled>> _lookupByTimeout;
+
+    // Functions to call on each item when inserting or removing from the queue.
+    function<void(T&)> _startFunction;
+    function<void(T&)> _endFunction;
+};
+
+template<typename T>
+void SScheduledPriorityQueue<T>::clear()  {
+    SAUTOLOCK(_queueMutex);
```

Comment:
nab because i know this was moved from the old implementation, but we've been moving away from `SAUTOLOCK`